### PR TITLE
safeStringLower and safeStringUpper convert defaultValue to lowercase and uppercase

### PR DESCRIPTION
**Pull request by samgermain**
# Updates

- for `safeStringUpper(dict, 'value', 'default')`, the default value is returned as `DEFAULT` and vice versa for `safeStringLower`

-------------

# Testing

# JavaScript 

## dict = {"test":"","test2":"10"}

`safe_string_lower (dict, 'test')`: `undefined`
`safe_string_upper (dict, 'test')`: `undefined`
`safe_string_lower_2 (dict, 'test', 'test2')`: `10`
`safe_string_upper_2 (dict, 'test', 'test2')`: `10`
`safe_string_lower_n (dict, ['test', 'test2'])`: `10`
`safe_string_upper_n (dict, ['test', 'test2'])`: `10`
`safe_string_lower (dict, 'missing_key', 'DEFAULT') *returns default*`: `default`
`safe_string_upper (dict, 'missing_key', 'default') *returns DEFAULT*`: `DEFAULT`
`safe_string_lower_2 (dict, 'missing_key', 'missing_key_2', 'DEFAULT') *returns default*`: `default`
`safe_string_upper_2 (dict, 'missing_key', 'missing_key_2', 'default') *returns DEFAULT*`: `DEFAULT`
`safe_string_lower_n (dict, ['missing_key', 'missing_key_2'], 'DEFAULT') *returns default*`: `default`
`safe_string_upper_n (dict, ['missing_key', 'missing_key_2'], 'default') *returns DEFAULT*`: `DEFAULT`

## dict = {"test":"5","test2":""}

`safe_string_lower (dict, 'test')`: `5`
`safe_string_upper (dict, 'test')`: `5`
`safe_string_lower_2 (dict, 'test', 'test2')`: `5`
`safe_string_upper_2 (dict, 'test', 'test2')`: `5`
`safe_string_lower_n (dict, ['test', 'test2'])`: `5`
`safe_string_upper_n (dict, ['test', 'test2'])`: `5`

## dict = {"test":"5","test2":"10"}

`safe_string_lower (dict, 'test')`: `5`
`safe_string_upper (dict, 'test')`: `5`
`safe_string_lower_2 (dict, 'test', 'test2')`: `5`
`safe_string_upper_2 (dict, 'test', 'test2')`: `5`
`safe_string_lower_n (dict, ['test', 'test2'])`: `5`
`safe_string_upper_n (dict, ['test', 'test2'])`: `5`

## dict = {"test":5}

`safe_string_lower (dict, 'test')`: `5`
`safe_string_upper (dict, 'test')`: `5`
`safe_string_lower_2 (dict, 'test', 'test2')`: `5`
`safe_string_upper_2 (dict, 'test', 'test2')`: `5`
`safe_string_lower_n (dict, ['test', 'test2'])`: `5`
`safe_string_upper_n (dict, ['test', 'test2'])`: `5`

## dict = {"test2":"5"}

`safe_string_lower (dict, 'test')`: `undefined`
`safe_string_upper (dict, 'test')`: `undefined`
`safe_string_lower_2 (dict, 'test', 'test2')`: `5`
`safe_string_upper_2 (dict, 'test', 'test2')`: `5`
`safe_string_lower_n (dict, ['test', 'test2'])`: `5`
`safe_string_upper_n (dict, ['test', 'test2'])`: `5`

## dict = {"test":""}

`safe_string_lower (dict, 'test')`: `undefined`
`safe_string_upper (dict, 'test')`: `undefined`
`safe_string_lower_2 (dict, 'test', 'test2')`: `undefined`
`safe_string_upper_2 (dict, 'test', 'test2')`: `undefined`
`safe_string_lower_n (dict, ['test', 'test2'])`: `undefined`
`safe_string_upper_n (dict, ['test', 'test2'])`: `undefined`

## dict = {"test2":""}

`safe_string_lower (dict, 'test')`: `undefined`
`safe_string_upper (dict, 'test')`: `undefined`
`safe_string_lower_2 (dict, 'test', 'test2')`: `undefined`
`safe_string_upper_2 (dict, 'test', 'test2')`: `undefined`
`safe_string_lower_n (dict, ['test', 'test2'])`: `undefined`
`safe_string_upper_n (dict, ['test', 'test2'])`: `undefined`

# python

safe_string_lower ('missing_key', 'DEFAULT') *returns default*: default
safe_string_upper ('missing_key', 'default') *returns DEFAULT*: DEFAULT
safe_string_lower_2 ('missing_key', 'missing_key_2', 'DEFAULT') *returns default*: default
safe_string_upper_2 ('missing_key', 'missing_key_2', 'default') *returns DEFAULT*: DEFAULT
safe_string_lower_n (['missing_key', 'missing_key_2'], 'DEFAULT') *returns default*: default
safe_string_upper_n (['missing_key', 'missing_key_2'], 'default') *returns DEFAULT*: DEFAULT

### {'test': '', 'test2': '10'}

safe_string_lower ('test'): None
safe_string_upper ('test'): None
safe_string_lower_2 ('test', 'test2'): 10
safe_string_upper_2 ('test', 'test2'): 10
safe_string_lower_n (['test', 'test2']): 
safe_string_upper_n (['test', 'test2']): 

### {'test': '5', 'test2': ''}

safe_string_lower ('test'): 5
safe_string_upper ('test'): 5
safe_string_lower_2 ('test', 'test2'): 5
safe_string_upper_2 ('test', 'test2'): 5
safe_string_lower_n (['test', 'test2']): 5
safe_string_upper_n (['test', 'test2']): 5
### {'test': '5', 'test2': '10'}

safe_string_lower ('test'): 5
safe_string_upper ('test'): 5
safe_string_lower_2 ('test', 'test2'): 5
safe_string_upper_2 ('test', 'test2'): 5
safe_string_lower_n (['test', 'test2']): 5
safe_string_upper_n (['test', 'test2']): 5

### {'test': 5}

safe_string_lower ('test'): 5
safe_string_upper ('test'): 5
safe_string_lower_2 ('test', 'test2'): 5
safe_string_upper_2 ('test', 'test2'): 5
safe_string_lower_n (['test', 'test2']): 5
safe_string_upper_n (['test', 'test2']): 5

### {'test2': '5'}

safe_string_lower ('test'): None
safe_string_upper ('test'): None
safe_string_lower_2 ('test', 'test2'): 5
safe_string_upper_2 ('test', 'test2'): 5
safe_string_lower_n (['test', 'test2']): 5
safe_string_upper_n (['test', 'test2']): 5

### {'test': ''}

safe_string_lower ('test'): None
safe_string_upper ('test'): None
safe_string_lower_2 ('test', 'test2'): None
safe_string_upper_2 ('test', 'test2'): None
safe_string_lower_n (['test', 'test2']): 
safe_string_upper_n (['test', 'test2']): 

### {'test2': ''}

safe_string_lower ('test'): None
safe_string_upper ('test'): None
safe_string_lower_2 ('test', 'test2'): None
safe_string_upper_2 ('test', 'test2'): None
safe_string_lower_n (['test', 'test2']): 
safe_string_upper_n (['test', 'test2']): 

# PHP

safe_string_lower ('missing_key', 'DEFAULT') *returns default*: default
safe_string_upper ('missing_key', 'default') *returns DEFAULT*: DEFAULT
safe_string_lower_2 ('missing_key', 'missing_key_2', 'DEFAULT') *returns default*: default
safe_string_upper_2 ('missing_key', 'missing_key_2', 'default') *returns DEFAULT*: DEFAULT
safe_string_lower_n (['missing_key', 'missing_key_2'], 'DEFAULT') *returns default*: default
safe_string_upper_n (['missing_key', 'missing_key_2'], 'default') *returns DEFAULT*: DEFAULT

### {"test":"","test2":"10"}

safe_string_lower ('test'): 
safe_string_upper ('test'): 
safe_string_lower_2 ('test', 'test2'): 10
safe_string_upper_2 ('test', 'test2'): 10
safe_string_lower_n (['test', 'test2']): 
safe_string_upper_n (['test', 'test2']): 

### {"test":"5","test2":"10"}

safe_string_lower ('test'): 5
safe_string_upper ('test'): 5
safe_string_lower_2 ('test', 'test2'): 5
safe_string_upper_2 ('test', 'test2'): 5
safe_string_lower_n (['test', 'test2']): 5
safe_string_upper_n (['test', 'test2']): 5

### {"test":"5","test2":"10"}

safe_string_lower ('test'): 5
safe_string_upper ('test'): 5
safe_string_lower_2 ('test', 'test2'): 5
safe_string_upper_2 ('test', 'test2'): 5
safe_string_lower_n (['test', 'test2']): 5
safe_string_upper_n (['test', 'test2']): 5

### {"test":"5"}

safe_string_lower ('test'): 5
safe_string_upper ('test'): 5
safe_string_lower_2 ('test', 'test2'): 5
safe_string_upper_2 ('test', 'test2'): 5
safe_string_lower_n (['test', 'test2']): 5
safe_string_upper_n (['test', 'test2']): 5

### {"test2":"5"}

safe_string_lower ('test'): 
safe_string_upper ('test'): 
safe_string_lower_2 ('test', 'test2'): 5
safe_string_upper_2 ('test', 'test2'): 5
safe_string_lower_n (['test', 'test2']): 5
safe_string_upper_n (['test', 'test2']): 5

### {"test":""}

safe_string_lower ('test'): 
safe_string_upper ('test'): 
safe_string_lower_2 ('test', 'test2'): 
safe_string_upper_2 ('test', 'test2'): 
safe_string_lower_n (['test', 'test2']): 
safe_string_upper_n (['test', 'test2']): 

### {"test2":""}

safe_string_lower ('test'): 
safe_string_upper ('test'): 
safe_string_lower_2 ('test', 'test2'): 
safe_string_upper_2 ('test', 'test2'): 
safe_string_lower_n (['test', 'test2']): 
safe_string_upper_n (['test', 'test2']):

### CODE REVIEW

File: python/ccxt/base/exchange.py [hunk 1]
```diff
@@ -853,7 +869,7 @@ def safe_value_n(dictionary, key_list, default_value=None):
 
     @staticmethod
     def get_object_value_from_key_list(dictionary, key_list):
-        filtered_list = list(filter(lambda el: el in dictionary, key_list))
+        filtered_list = list(filter(lambda el: el in dictionary and dictionary[el] != '' and dictionary[el] is not None, key_list))
```
Comment (carlosmiei) [inline python/ccxt/base/exchange.py:872]: @samgermain why is this needed? can you give us a real-case scenario?

Comment (samgermain) [reply python/ccxt/base/exchange.py:872]: These are the examples, I'll need to go to a PR I did about 6 months ago to find the real use case, I fixed the problem for `safe_string_2` but not `safe_string_n`

```python
safe_string_lower_n ({'test': '', 'test2': '10'}, ['test', 'test2'])    # should return 10, actually returns ''
safe_string_upper_n ({'test': '', 'test2': '10'}, ['test', 'test2'])    # should return 10, actually returns ''
safe_string_lower_n ({'test': ''}, ['test', 'test2'])                   # should return None, actually returns ''
safe_string_upper_n ({'test': ''}, ['test', 'test2'])                   # should return None, actually returns ''
safe_string_lower_n ({'test2': ''}, ['test', 'test2'])                  # should return None, actually returns ''
safe_string_upper_n ({'test2': ''}, ['test', 'test2'])                  # should return None, actually returns ''
```

Comment (samgermain) [reply python/ccxt/base/exchange.py:872]: [Here's the commit where it was fixed for `safeString2`](https://github.com/ccxt/ccxt/pull/13438)

It looks like [OKX stop orders were having issues with dictionary/object values being empty strings](https://github.com/ccxt/ccxt/pull/13440)

Comment (carlosmiei) [reply python/ccxt/base/exchange.py:872]: @samgermain Yeah okx might have a weird api but I'm not sure we should do that by default.

For instance:
```Javascript
safe_string_lower_n ({'test': ''}, ['test', 'test2'], "defaultValue") // will return "defaultValue"
```   
⬆️ why should we return the defaultValue here and hide the key existence?

Comment (samgermain) [reply python/ccxt/base/exchange.py:872]: > @samgermain Yeah okx might have a weird api but I'm not sure we should do that by default.
> 
> For instance:
> 
> ```js
> safe_string_lower_n ({'test': ''}, ['test', 'test2'], "defaultValue") // will return "defaultValue"
> ```
> 
> ⬆️ why should we return the defaultValue here and hide the key existence?

Because there isn't a value, if we need a number for string math or something, we would need to perform the math on either the value that the exchange gives us, or on the default value, but we don't want to perform it on an empty string, which is the absence of a value

Comment (samgermain) [reply python/ccxt/base/exchange.py]: I moved this change to a separate PR https://github.com/ccxt/ccxt/pull/15854

Comment (carlosmiei) [reply python/ccxt/base/exchange.py]: @samgermain thanks we can keep this discussion there then.

**Review by carlosmiei — APPROVED** [review]
